src/config.rs

```diff
@@ -110,6 +116,112 @@ configuration_option_enum! { TypeDensity:
     Wide,
 }
 
+// Newtype needed to have impls for `create_config` macro.
+#[derive(Clone, Debug, Default, RustcDecodable)]
+pub struct FileLinesMap(pub HashMap<String, LineSet>);
+
+impl ConfigType for Option<FileLinesMap> {
+    type ParseErr = String;
+
+    fn doc_hint() -> String {
+        unimplemented!();
+    }
+
+    fn parse(s: &str) -> Result<Self, String> {
+        let mut map = HashMap::new();
+        let (file, lines) = try!(parse_file_lines_spec(s));
+        map.insert(file, lines);
+
+        Ok(Some(FileLinesMap(map)))
+    }
+}
+
+pub fn parse_file_lines_spec(s: &str) -> Result<(String, LineSet), String> {
+    let err = || Err(format!("invalid experimental-file-lines argument: {}", s));
+
+    match file_lines_spec_parser::file_lines_spec(&s) {
+        IResult::Error(_) |
+        IResult::Incomplete(_) => return err(),
+        IResult::Done(remaining, _) if !remaining.is_empty() => return err(),
+        IResult::Done(_, (file, line_ranges)) => Ok((file, line_ranges)),
+    }
+}
+
+impl iter::FromIterator<(String, LineSet)> for FileLinesMap {
+    fn from_iter<I: IntoIterator<Item = (String, LineSet)>>(iter: I) -> FileLinesMap {
+        let mut ret = FileLinesMap(HashMap::new());
+        ret.extend(iter);
+        ret
+    }
+}
+
+impl Extend<(String, LineSet)> for FileLinesMap {
+    fn extend<I>(&mut self, iter: I)
+        where I: IntoIterator<Item = (String, LineSet)>
+    {
+        let map = &mut self.0;
+
+        for (file, line_set) in iter {
+            map.entry(file).or_insert(LineSet::new()).extend(line_set);
+        }
+    }
+}
+
+mod file_lines_spec_parser {
```

Comment:
> Rather than all this parsing code, could we just deserialise a JSON string to some structs?

I'm kind of against JSON in command line arguments in general, but I'll see what that looks like.

> Could also use an empty array to mean no spec, rather than using an Option, which should mean fewer changes to the config code.

You mean empty map? Will give that a shot as well.